cmd/yggd/worker.go

```diff
@@ -65,27 +278,52 @@ func watchWorkerDir(dir string, env []string, died chan int) {
 		log.Debugf("received inotify event %v", e.Event())
 		switch e.Event() {
 		case notify.InCloseWrite, notify.InMovedTo:
-			if strings.HasSuffix(e.Path(), "worker") {
-				if ExcludeWorkers[filepath.Base(e.Path())] {
+			name := filepath.Base(e.Path())
+			if validWorker(name) {
+				var config *workerConfig
+
+				configFile := filepath.Join(workerConfigDir, name+".toml")
+				if fileExists(configFile) {
+					var err error
+
+					config, err = loadWorkerConfig(configFile)
+					if err != nil {
+						log.Errorf("cannot read worker config file: %v", err)
+						continue
+					}
+				} else {
+					config = &workerConfig{}
+				}
+				config.program = filepath.Join(workerExecDir, name)
+
+				if err := startWorker(*config, nil, func(pid int) {
+					died <- pid
+				}); err != nil {
+					log.Errorf("cannot start worker %v: %v", config.program, err)
 					continue
 				}
-				log.Tracef("new worker detected: %v", e.Path())
-				go startProcess(e.Path(), env, 0, died)
 			}
 		case notify.InDelete, notify.InMovedFrom:
 			workerName := filepath.Base(e.Path())
-			pidFilePath := filepath.Join(
-				yggdrasil.LocalstateDir,
-				"run",
-				yggdrasil.LongName,
-				"workers",
-				workerName+".pid",
-			)
-
-			if err := killWorker(pidFilePath); err != nil {
-				log.Errorf("cannot kill worker: %v", err)
+			if err := stopWorker(workerName); err != nil {
+				log.Errorf("cannot stop worker: %v", err)
 				continue
 			}
 		}
 	}
 }
+
+func validEnvVar(val string) bool {
+	for _, pattern := range []string{"PATH=.*", "YGG_.*=.*"} {
```

Comment:
What about `LD_PRELOAD` or `PYTHONPATH`?

Reply:
Ah, yea, those are probably worth keeping out, but if we start to exclude every language's library search path, where will it end? Maybe it's better to prefix each inserted variable with the name of the worker, thereby permitting all variables, but with running the risk of clobbering conventional ones like `LD_PRELOAD`?

Reply:
Oh. That's an idea. Are you suggesting a per-worker list of forbidden environment variables? Or a single, global list of forbidden environment variables that's kept in `/etc/rhc/config.toml`?

Reply:
I think that It would make sense to have both of them. One global list containing e.g. `PATH`, `LD_PRELOAD`, `YGG_.*` environment variables, and then per-worker list. When worker is written in Python, then it makes sense to extend global list with `PYTHONPATH` from per-worker list for given worker. It would have less sense to have such variable for worker written in Go or C.

On the other side, I'm not sure if it is worth adding such new complexity to this branch?

Reply:
Yea. Since this isn't a configuration that's going to be carried forward into yggdrasil 0.3+, I'd rather not add a configuration value that we aren't supporting in the future. Can we leave it as a hard-coded list of `PATH` and `YGG_*` and call that good enough for now?

Reply:
Yeah, I think there is small risk from environment variables, because yggdrasil is typically started using systemd service. Only root can modify environment variables in this service.